{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: MacOSX\n",
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/IPython/core/magics/pylab.py:159: UserWarning: pylab import has clobbered these variables: ['median', 'percentile']\n",
      "`%matplotlib` prevents importing * from pylab and numpy\n",
      "  warn(\"pylab import has clobbered these variables: %s\"  % clobbered +\n"
     ]
    }
   ],
   "source": [
    "%pylab\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder\n",
    "from sklearn.cross_decomposition import PLSRegression\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.stats import zscore\n",
    "import mplcursors\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "DeSeqOutputAllConds = 'DeSeqOutputAllConds.tsv'\n",
    "si2_si4_RNA_seq = 'si2-si4_RNA-seq-pipeline-output-normalized.tsv'\n",
    "le = LabelEncoder()\n",
    "oh_enc = OneHotEncoder(handle_unknown='ignore')\n",
    "plsr = PLSRegression(n_components=2)\n",
    "\n",
    "#https://stackoverflow.com/questions/11346283/renaming-columns-in-pandas\n",
    "def combineTreatmentColumns(df, exceptions):\n",
    "    df.rename(columns=lambda x: x[3:] if x not in exceptions else x, inplace=True)\n",
    "    return df\n",
    "\n",
    "def changeTreatmentNameInRows(df, columnName='sampleID', exceptions={}):\n",
    "    df[columnName] = df[columnName].apply(lambda x: x[3:] if x not in exceptions else x)\n",
    "    return df\n",
    "\n",
    "\n",
    "def getTSVFile(filepath):\n",
    "    data = pd.read_csv (filepath, sep = '\\t')\n",
    "    return data\n",
    "\n",
    "#not used or necessary\n",
    "def integerEncode(dataframe, columnNametoEncode, newEncodedColumnName):\n",
    "    dataframe[columnNametoEncode] = le.fit_transform(df[newEncodedColumnName])\n",
    "    return dataframe\n",
    "  \n",
    "#https://stackoverflow.com/questions/37292872/how-can-i-one-hot-encode-in-python  \n",
    "def oneHotEncode(df, columnsToEncode):\n",
    "    \n",
    "    for column in columnsToEncode:\n",
    "        # Get one hot encoding of column and drop 'column'\n",
    "        one_hot = pd.get_dummies(df[column])\n",
    "        # Drop column  as it is now encoded\n",
    "        df = df.drop(column, axis = 1)\n",
    "        # Join the encoded df\n",
    "        df = df.join(one_hot)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def zScoreData(df, columnsToZScore):\n",
    "    for column in columnsToZScore:\n",
    "        df[column] = zscore(df[column])\n",
    "        \n",
    "    return df\n",
    "\n",
    "def fixBadChromosomeLabeling(df, column='chrom'): \n",
    "    replace = list(range(1, 23)) + ['X', 'Y']\n",
    "    \n",
    "    for i in replace:\n",
    "        df[column].replace(str(i), 'chr' + str(i), inplace=True)\n",
    "        \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = getTSVFile(si2_si4_RNA_seq)\n",
    "df = df.loc[df['counts'] > 0]#Filter out rows where count is zero\n",
    "df = fixBadChromosomeLabeling(df)\n",
    "\n",
    "#Set up copies to play around with lader\n",
    "df_original = df.copy()\n",
    "\n",
    "#Remove unnecessary columns\n",
    "df = df[['sampleID', 'tpm', 'gene_name']]\n",
    "\n",
    "#display(df.sampleID.unique())\n",
    "\n",
    "df = oneHotEncode(df, ['sampleID'])\n",
    "df = zScoreData(df, ['tpm'])\n",
    "\n",
    "\n",
    "# Leon if you want to get top n%\n",
    "# Probably simpler - https://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.DataFrame.nlargest.html\n",
    "# Sort, .head(...) - https://stackoverflow.com/questions/53405458/selecting-the-top-50-percentage-names-from-the-columns-of-a-pandas-dataframe\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22112\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>counts</th>\n",
       "      <th>rpm</th>\n",
       "      <th>rpkm</th>\n",
       "      <th>tpm</th>\n",
       "      <th>tx_start</th>\n",
       "      <th>tx_end</th>\n",
       "      <th>TSS_loc</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RA-high</th>\n",
       "      <td>609.162507</td>\n",
       "      <td>60.336679</td>\n",
       "      <td>16.533681</td>\n",
       "      <td>60.336679</td>\n",
       "      <td>7.337781e+07</td>\n",
       "      <td>7.343661e+07</td>\n",
       "      <td>7.340665e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EtOH-nlDensity</th>\n",
       "      <td>598.806172</td>\n",
       "      <td>60.246408</td>\n",
       "      <td>16.328527</td>\n",
       "      <td>60.246408</td>\n",
       "      <td>7.338924e+07</td>\n",
       "      <td>7.344843e+07</td>\n",
       "      <td>7.341827e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA-med</th>\n",
       "      <td>605.192744</td>\n",
       "      <td>60.195032</td>\n",
       "      <td>16.438527</td>\n",
       "      <td>60.195032</td>\n",
       "      <td>7.350551e+07</td>\n",
       "      <td>7.356460e+07</td>\n",
       "      <td>7.353451e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EtOH-halfDensity</th>\n",
       "      <td>594.354948</td>\n",
       "      <td>60.042029</td>\n",
       "      <td>16.434342</td>\n",
       "      <td>60.042029</td>\n",
       "      <td>7.341303e+07</td>\n",
       "      <td>7.347192e+07</td>\n",
       "      <td>7.344199e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RA-low</th>\n",
       "      <td>632.429553</td>\n",
       "      <td>59.912528</td>\n",
       "      <td>16.294389</td>\n",
       "      <td>59.912528</td>\n",
       "      <td>7.345944e+07</td>\n",
       "      <td>7.351828e+07</td>\n",
       "      <td>7.348816e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EtOH-highDensity</th>\n",
       "      <td>588.139612</td>\n",
       "      <td>59.885021</td>\n",
       "      <td>16.443999</td>\n",
       "      <td>59.885021</td>\n",
       "      <td>7.337178e+07</td>\n",
       "      <td>7.343102e+07</td>\n",
       "      <td>7.340086e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFb-and-RA-low</th>\n",
       "      <td>595.760855</td>\n",
       "      <td>59.861122</td>\n",
       "      <td>15.898275</td>\n",
       "      <td>59.861122</td>\n",
       "      <td>7.352884e+07</td>\n",
       "      <td>7.358795e+07</td>\n",
       "      <td>7.355775e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFb-and-RA-high</th>\n",
       "      <td>576.474598</td>\n",
       "      <td>59.675366</td>\n",
       "      <td>16.140550</td>\n",
       "      <td>59.675366</td>\n",
       "      <td>7.352537e+07</td>\n",
       "      <td>7.358423e+07</td>\n",
       "      <td>7.355415e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFb-and-RA-med</th>\n",
       "      <td>633.242105</td>\n",
       "      <td>59.509641</td>\n",
       "      <td>15.752275</td>\n",
       "      <td>59.509641</td>\n",
       "      <td>7.358892e+07</td>\n",
       "      <td>7.364770e+07</td>\n",
       "      <td>7.361773e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFb-high</th>\n",
       "      <td>623.298412</td>\n",
       "      <td>59.468353</td>\n",
       "      <td>16.066546</td>\n",
       "      <td>59.468353</td>\n",
       "      <td>7.364203e+07</td>\n",
       "      <td>7.370125e+07</td>\n",
       "      <td>7.367105e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFb-low</th>\n",
       "      <td>642.757235</td>\n",
       "      <td>59.428300</td>\n",
       "      <td>15.976519</td>\n",
       "      <td>59.428300</td>\n",
       "      <td>7.346285e+07</td>\n",
       "      <td>7.352204e+07</td>\n",
       "      <td>7.349189e+07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGFb-med</th>\n",
       "      <td>632.534628</td>\n",
       "      <td>59.277994</td>\n",
       "      <td>15.837809</td>\n",
       "      <td>59.277994</td>\n",
       "      <td>7.360769e+07</td>\n",
       "      <td>7.366711e+07</td>\n",
       "      <td>7.363677e+07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      counts        rpm       rpkm        tpm      tx_start  \\\n",
       "sampleID                                                                      \n",
       "RA-high           609.162507  60.336679  16.533681  60.336679  7.337781e+07   \n",
       "EtOH-nlDensity    598.806172  60.246408  16.328527  60.246408  7.338924e+07   \n",
       "RA-med            605.192744  60.195032  16.438527  60.195032  7.350551e+07   \n",
       "EtOH-halfDensity  594.354948  60.042029  16.434342  60.042029  7.341303e+07   \n",
       "RA-low            632.429553  59.912528  16.294389  59.912528  7.345944e+07   \n",
       "EtOH-highDensity  588.139612  59.885021  16.443999  59.885021  7.337178e+07   \n",
       "TGFb-and-RA-low   595.760855  59.861122  15.898275  59.861122  7.352884e+07   \n",
       "TGFb-and-RA-high  576.474598  59.675366  16.140550  59.675366  7.352537e+07   \n",
       "TGFb-and-RA-med   633.242105  59.509641  15.752275  59.509641  7.358892e+07   \n",
       "TGFb-high         623.298412  59.468353  16.066546  59.468353  7.364203e+07   \n",
       "TGFb-low          642.757235  59.428300  15.976519  59.428300  7.346285e+07   \n",
       "TGFb-med          632.534628  59.277994  15.837809  59.277994  7.360769e+07   \n",
       "\n",
       "                        tx_end       TSS_loc  \n",
       "sampleID                                      \n",
       "RA-high           7.343661e+07  7.340665e+07  \n",
       "EtOH-nlDensity    7.344843e+07  7.341827e+07  \n",
       "RA-med            7.356460e+07  7.353451e+07  \n",
       "EtOH-halfDensity  7.347192e+07  7.344199e+07  \n",
       "RA-low            7.351828e+07  7.348816e+07  \n",
       "EtOH-highDensity  7.343102e+07  7.340086e+07  \n",
       "TGFb-and-RA-low   7.358795e+07  7.355775e+07  \n",
       "TGFb-and-RA-high  7.358423e+07  7.355415e+07  \n",
       "TGFb-and-RA-med   7.364770e+07  7.361773e+07  \n",
       "TGFb-high         7.370125e+07  7.367105e+07  \n",
       "TGFb-low          7.352204e+07  7.349189e+07  \n",
       "TGFb-med          7.366711e+07  7.363677e+07  "
      ]
     },
     "execution_count": 357,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Just a cell for testing the \"changeTreatmentNameInRows\" function which \n",
    "#removes the number '31-' from '31-RA-high' and groups them to see new values\n",
    "\n",
    "df_test = df_original.copy()\n",
    "df_test = df_test.loc[df_test['counts'] > 0]\n",
    "\n",
    "changeTreatmentNameInRows(df_test)\n",
    "\n",
    "#Homework: Get top 10 results, for gene_name, sampleID, and both and PLOT IT\n",
    "df_test.groupby(['sampleID' ]).mean().sort_values(by=['tpm'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retinoic Acid (50, 200 and 400 nM)\n",
    "# TGF-β (1.25, 5, and 10 ng/mL)\n",
    "# EtOH - Ethanol control\n",
    "\n",
    "#Transforms columns with these keys to columns with these values\n",
    "sampleIDs_transform = {\n",
    "  'RA-high': {'TGFb': 0, 'retinoicAcid': 400, 'EtOH': 0}, \n",
    "  'EtOH-nlDensity': {'TGFb': 0, 'retinoicAcid': 0, 'EtOH': 0},\n",
    "  'RA-med': {'TGFb': 0, 'retinoicAcid': 200, 'EtOH': 0}, \n",
    "  'EtOH-halfDensity': {'TGFb': 0, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'RA-low': {'TGFb': 0, 'retinoicAcid': 50, 'EtOH': 0},\n",
    "  'EtOH-highDensity': {'TGFb': 0, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'TGFb-and-RA-low': {'TGFb': 1.25, 'retinoicAcid': 50, 'EtOH': 0}, \n",
    "  'TGFb-and-RA-high': {'TGFb': 10, 'retinoicAcid': 400, 'EtOH': 0},\n",
    "  'TGFb-and-RA-med': {'TGFb': 5, 'retinoicAcid': 200, 'EtOH': 0}, \n",
    "  'TGFb-high': {'TGFb': 10, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'TGFb-low': {'TGFb': 1.25, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'TGFb-med': {'TGFb': 5, 'retinoicAcid': 0, 'EtOH': 0}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLSR Variance captured:  0.0009713932769096489\n"
     ]
    }
   ],
   "source": [
    "\n",
    "    \n",
    "#Simply takes the X and Y loadings and plots them with their legend\n",
    "def printPLSRPlot(x_loadings_, y_loadings_, loadings):\n",
    "    plt.scatter(y_loadings_[:, 0], y_loadings_[:, 1], color='blue', label='tpm')\n",
    "    \n",
    "    for i in range(x_loadings_.shape[0]):\n",
    "        plt.scatter(x_loadings_[i,0], x_loadings_[i, 1], label=loadings[i])\n",
    "        \n",
    "    plt.legend(bbox_to_anchor=(1.53, 1))\n",
    "    plt.xlabel('PC1')\n",
    "    plt.ylabel('PC2')\n",
    "    plt.show()\n",
    "  \n",
    "#Creates a new datatable by grouping by two columns then returning a new \n",
    "#datatable where the rows were the first groupby element and the columns were the 2nd groupby\n",
    "def getPLSR_DFs(groupby=['sampleID', 'gene_name'], columns=['sampleID', 'tpm', 'gene_name'], topN=5000):\n",
    "    df_plsr = df_original.copy()\n",
    "    df_plsr = df_plsr[columns]\n",
    "\n",
    "    gb = df_plsr.groupby(groupby).mean().sort_values(by=['tpm'], ascending=False)[:topN]\n",
    "    #Extract rows and columns of top N results by indices\n",
    "    rows = set([groupby[0] for groupby in gb.index])\n",
    "    columns = set([groupby[1] for groupby in gb.index])\n",
    "\n",
    "    #Create a new dataframe\n",
    "    new_df = pd.DataFrame(0, index=rows, columns=columns)\n",
    "    new_df_log = new_df.copy()\n",
    "\n",
    "    #Iterate\n",
    "    for (row, col) in gb.index:\n",
    "        tpv_val = gb.loc[row, col].tpm\n",
    "        new_df.loc[row, col] = tpv_val\n",
    "        new_df_log.loc[row, col] = np.log(tpv_val)\n",
    "        \n",
    "    return new_df, new_df_log\n",
    "\n",
    "#Uses \"sampleIDs_transform\" to put in the correct treatments in their \n",
    "#appropriate columns\n",
    "def alterTreatmentsInDF(new_df, df, newColumns=['TGFb', 'retinoicAcid']):\n",
    "    \n",
    "    for i in range(len(new_df.index)):\n",
    "        \n",
    "        (gene_name, treatment) = df.iloc[i].name\n",
    "        new_df.iat[i, 0] = sampleIDs_transform[treatment]['TGFb']\n",
    "        new_df.iat[i, 1]= sampleIDs_transform[treatment]['retinoicAcid']\n",
    "        #new_df.iat[i, 2]= sampleIDs_transform[treatment]['EtOH']\n",
    "        \n",
    "    return new_df\n",
    "\n",
    "#Takes a multiindex dataframe, creates a new dataframe, and inputs the correct \n",
    "#treatments into it\n",
    "def createPLSRMatrix(df, newColumns=['TGFb', 'retinoicAcid']):\n",
    "    #Create a new dataframe    \n",
    "    new_indexes = [gene for (gene, treatment) in df.index]\n",
    "    new_df = pd.DataFrame(0, index=new_indexes, columns=newColumns)\n",
    "            \n",
    "    return alterTreatmentsInDF(new_df, df)\n",
    "\n",
    "#In this section Im just fitting the new dataframe with columns \n",
    "#\"TGF-b\" and \"Retinoic Acid\" to results\n",
    "df_plsr = df_original.copy()\n",
    "df_plsr = df_plsr[['sampleID', 'tpm', 'gene_name']]\n",
    "changeTreatmentNameInRows(df_plsr, columnName='sampleID')\n",
    "\n",
    "\n",
    "df_plsr = df_plsr.groupby(['gene_name', 'sampleID']).mean().sort_values(by=['tpm'], ascending=False)[:50]\n",
    "Y = df_plsr['tpm']\n",
    "\n",
    "\n",
    "better_df = createPLSRMatrix(df_plsr)\n",
    "\n",
    "loadings = [col for col in df_plsr.columns if col != 'tpm'] #Remove 'tpm' to fit for PLSR\n",
    "plsr.fit(better_df, Y)\n",
    "print('PLSR Variance captured: ', plsr.score(better_df, Y))\n",
    "printPLSRPlot(plsr.x_loadings_, plsr.y_loadings_, better_df.columns)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>TGFb</th>\n",
       "      <th>Retinoic Acid</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.377997</td>\n",
       "      <td>0.035485</td>\n",
       "      <td>0.087716</td>\n",
       "      <td>0.263830</td>\n",
       "      <td>0.350411</td>\n",
       "      <td>12.057979</td>\n",
       "      <td>0.319075</td>\n",
       "      <td>53.020671</td>\n",
       "      <td>...</td>\n",
       "      <td>37.376160</td>\n",
       "      <td>182.345536</td>\n",
       "      <td>3.376170</td>\n",
       "      <td>9.657581</td>\n",
       "      <td>36.781075</td>\n",
       "      <td>16.005638</td>\n",
       "      <td>18.488987</td>\n",
       "      <td>72.885622</td>\n",
       "      <td>33.661747</td>\n",
       "      <td>22.205224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.00</td>\n",
       "      <td>50</td>\n",
       "      <td>0.345151</td>\n",
       "      <td>0.034716</td>\n",
       "      <td>0.127103</td>\n",
       "      <td>0.137183</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.425112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>47.729696</td>\n",
       "      <td>...</td>\n",
       "      <td>37.457048</td>\n",
       "      <td>150.343398</td>\n",
       "      <td>2.518407</td>\n",
       "      <td>9.277073</td>\n",
       "      <td>33.313392</td>\n",
       "      <td>16.573176</td>\n",
       "      <td>20.549266</td>\n",
       "      <td>107.455344</td>\n",
       "      <td>33.147009</td>\n",
       "      <td>23.398570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.00</td>\n",
       "      <td>200</td>\n",
       "      <td>0.300917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.162882</td>\n",
       "      <td>0.082244</td>\n",
       "      <td>0.345012</td>\n",
       "      <td>13.311664</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>47.800100</td>\n",
       "      <td>...</td>\n",
       "      <td>30.569839</td>\n",
       "      <td>130.592939</td>\n",
       "      <td>2.773964</td>\n",
       "      <td>7.611645</td>\n",
       "      <td>33.531139</td>\n",
       "      <td>15.295295</td>\n",
       "      <td>20.328332</td>\n",
       "      <td>118.413674</td>\n",
       "      <td>32.668651</td>\n",
       "      <td>22.805652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.00</td>\n",
       "      <td>400</td>\n",
       "      <td>0.203307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.211807</td>\n",
       "      <td>0.069556</td>\n",
       "      <td>0.335723</td>\n",
       "      <td>13.785032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>46.066403</td>\n",
       "      <td>...</td>\n",
       "      <td>32.521534</td>\n",
       "      <td>123.678421</td>\n",
       "      <td>2.668080</td>\n",
       "      <td>8.206404</td>\n",
       "      <td>32.979449</td>\n",
       "      <td>15.033746</td>\n",
       "      <td>20.089239</td>\n",
       "      <td>121.144959</td>\n",
       "      <td>30.442728</td>\n",
       "      <td>23.403616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1.25</td>\n",
       "      <td>0</td>\n",
       "      <td>0.256296</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.080188</td>\n",
       "      <td>0.125608</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.649623</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>45.712356</td>\n",
       "      <td>...</td>\n",
       "      <td>27.967913</td>\n",
       "      <td>126.723619</td>\n",
       "      <td>3.286000</td>\n",
       "      <td>11.075037</td>\n",
       "      <td>37.483739</td>\n",
       "      <td>14.895510</td>\n",
       "      <td>17.376096</td>\n",
       "      <td>149.058857</td>\n",
       "      <td>32.772074</td>\n",
       "      <td>23.163784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1.25</td>\n",
       "      <td>50</td>\n",
       "      <td>0.351086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054925</td>\n",
       "      <td>0.075671</td>\n",
       "      <td>0.385303</td>\n",
       "      <td>8.973253</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>41.863051</td>\n",
       "      <td>...</td>\n",
       "      <td>27.021822</td>\n",
       "      <td>74.655933</td>\n",
       "      <td>3.736386</td>\n",
       "      <td>10.136228</td>\n",
       "      <td>37.666414</td>\n",
       "      <td>13.484046</td>\n",
       "      <td>21.128904</td>\n",
       "      <td>194.584197</td>\n",
       "      <td>32.281384</td>\n",
       "      <td>23.780587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.230624</td>\n",
       "      <td>0.035837</td>\n",
       "      <td>0.053812</td>\n",
       "      <td>0.110743</td>\n",
       "      <td>0.336408</td>\n",
       "      <td>9.336744</td>\n",
       "      <td>0.193978</td>\n",
       "      <td>43.406801</td>\n",
       "      <td>...</td>\n",
       "      <td>32.606337</td>\n",
       "      <td>110.171030</td>\n",
       "      <td>3.162182</td>\n",
       "      <td>11.979568</td>\n",
       "      <td>35.755054</td>\n",
       "      <td>14.319992</td>\n",
       "      <td>19.077618</td>\n",
       "      <td>153.719588</td>\n",
       "      <td>32.896717</td>\n",
       "      <td>23.078828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5.00</td>\n",
       "      <td>200</td>\n",
       "      <td>0.362810</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054184</td>\n",
       "      <td>0.050074</td>\n",
       "      <td>0.676265</td>\n",
       "      <td>9.562037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>42.264973</td>\n",
       "      <td>...</td>\n",
       "      <td>25.416376</td>\n",
       "      <td>66.817892</td>\n",
       "      <td>3.253249</td>\n",
       "      <td>10.199674</td>\n",
       "      <td>36.682645</td>\n",
       "      <td>13.111032</td>\n",
       "      <td>21.180285</td>\n",
       "      <td>207.092959</td>\n",
       "      <td>31.524216</td>\n",
       "      <td>24.633575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>10.00</td>\n",
       "      <td>0</td>\n",
       "      <td>0.350926</td>\n",
       "      <td>0.032057</td>\n",
       "      <td>0.092482</td>\n",
       "      <td>0.130880</td>\n",
       "      <td>0.358890</td>\n",
       "      <td>10.128333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>43.736571</td>\n",
       "      <td>...</td>\n",
       "      <td>28.349972</td>\n",
       "      <td>100.414759</td>\n",
       "      <td>3.536064</td>\n",
       "      <td>12.396366</td>\n",
       "      <td>36.297142</td>\n",
       "      <td>13.755184</td>\n",
       "      <td>18.309939</td>\n",
       "      <td>161.319258</td>\n",
       "      <td>32.275104</td>\n",
       "      <td>22.633430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10.00</td>\n",
       "      <td>400</td>\n",
       "      <td>0.419643</td>\n",
       "      <td>0.046901</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.098149</td>\n",
       "      <td>0.376772</td>\n",
       "      <td>10.329859</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40.635957</td>\n",
       "      <td>...</td>\n",
       "      <td>22.338756</td>\n",
       "      <td>57.523884</td>\n",
       "      <td>3.284246</td>\n",
       "      <td>9.952191</td>\n",
       "      <td>36.076187</td>\n",
       "      <td>13.478254</td>\n",
       "      <td>22.231341</td>\n",
       "      <td>204.279198</td>\n",
       "      <td>29.913855</td>\n",
       "      <td>25.225166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 22114 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name   TGFb  Retinoic Acid      A1BG      A1CF       A2M     A2ML1  \\\n",
       "0           0.00              0  0.377997  0.035485  0.087716  0.263830   \n",
       "1           0.00             50  0.345151  0.034716  0.127103  0.137183   \n",
       "2           0.00            200  0.300917  0.000000  0.162882  0.082244   \n",
       "3           0.00            400  0.203307  0.000000  0.211807  0.069556   \n",
       "4           1.25              0  0.256296  0.000000  0.080188  0.125608   \n",
       "5           1.25             50  0.351086  0.000000  0.054925  0.075671   \n",
       "6           5.00              0  0.230624  0.035837  0.053812  0.110743   \n",
       "7           5.00            200  0.362810  0.000000  0.054184  0.050074   \n",
       "8          10.00              0  0.350926  0.032057  0.092482  0.130880   \n",
       "9          10.00            400  0.419643  0.046901  0.000000  0.098149   \n",
       "\n",
       "gene_name   A3GALT2     A4GALT     A4GNT       AAAS  ...     ZWILCH  \\\n",
       "0          0.350411  12.057979  0.319075  53.020671  ...  37.376160   \n",
       "1          0.000000  13.425112  0.000000  47.729696  ...  37.457048   \n",
       "2          0.345012  13.311664  0.000000  47.800100  ...  30.569839   \n",
       "3          0.335723  13.785032  0.000000  46.066403  ...  32.521534   \n",
       "4          0.000000  11.649623  0.000000  45.712356  ...  27.967913   \n",
       "5          0.385303   8.973253  0.000000  41.863051  ...  27.021822   \n",
       "6          0.336408   9.336744  0.193978  43.406801  ...  32.606337   \n",
       "7          0.676265   9.562037  0.000000  42.264973  ...  25.416376   \n",
       "8          0.358890  10.128333  0.000000  43.736571  ...  28.349972   \n",
       "9          0.376772  10.329859  0.000000  40.635957  ...  22.338756   \n",
       "\n",
       "gene_name       ZWINT      ZXDA       ZXDB       ZXDC     ZYG11A     ZYG11B  \\\n",
       "0          182.345536  3.376170   9.657581  36.781075  16.005638  18.488987   \n",
       "1          150.343398  2.518407   9.277073  33.313392  16.573176  20.549266   \n",
       "2          130.592939  2.773964   7.611645  33.531139  15.295295  20.328332   \n",
       "3          123.678421  2.668080   8.206404  32.979449  15.033746  20.089239   \n",
       "4          126.723619  3.286000  11.075037  37.483739  14.895510  17.376096   \n",
       "5           74.655933  3.736386  10.136228  37.666414  13.484046  21.128904   \n",
       "6          110.171030  3.162182  11.979568  35.755054  14.319992  19.077618   \n",
       "7           66.817892  3.253249  10.199674  36.682645  13.111032  21.180285   \n",
       "8          100.414759  3.536064  12.396366  36.297142  13.755184  18.309939   \n",
       "9           57.523884  3.284246   9.952191  36.076187  13.478254  22.231341   \n",
       "\n",
       "gene_name         ZYX      ZZEF1       ZZZ3  \n",
       "0           72.885622  33.661747  22.205224  \n",
       "1          107.455344  33.147009  23.398570  \n",
       "2          118.413674  32.668651  22.805652  \n",
       "3          121.144959  30.442728  23.403616  \n",
       "4          149.058857  32.772074  23.163784  \n",
       "5          194.584197  32.281384  23.780587  \n",
       "6          153.719588  32.896717  23.078828  \n",
       "7          207.092959  31.524216  24.633575  \n",
       "8          161.319258  32.275104  22.633430  \n",
       "9          204.279198  29.913855  25.225166  \n",
       "\n",
       "[10 rows x 22114 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.377997</td>\n",
       "      <td>0.035485</td>\n",
       "      <td>0.087716</td>\n",
       "      <td>0.263830</td>\n",
       "      <td>0.350411</td>\n",
       "      <td>12.057979</td>\n",
       "      <td>0.319075</td>\n",
       "      <td>53.020671</td>\n",
       "      <td>20.904988</td>\n",
       "      <td>0.338778</td>\n",
       "      <td>...</td>\n",
       "      <td>37.376160</td>\n",
       "      <td>182.345536</td>\n",
       "      <td>3.376170</td>\n",
       "      <td>9.657581</td>\n",
       "      <td>36.781075</td>\n",
       "      <td>16.005638</td>\n",
       "      <td>18.488987</td>\n",
       "      <td>72.885622</td>\n",
       "      <td>33.661747</td>\n",
       "      <td>22.205224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.345151</td>\n",
       "      <td>0.034716</td>\n",
       "      <td>0.127103</td>\n",
       "      <td>0.137183</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.425112</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>47.729696</td>\n",
       "      <td>21.923315</td>\n",
       "      <td>0.227967</td>\n",
       "      <td>...</td>\n",
       "      <td>37.457048</td>\n",
       "      <td>150.343398</td>\n",
       "      <td>2.518407</td>\n",
       "      <td>9.277073</td>\n",
       "      <td>33.313392</td>\n",
       "      <td>16.573176</td>\n",
       "      <td>20.549266</td>\n",
       "      <td>107.455344</td>\n",
       "      <td>33.147009</td>\n",
       "      <td>23.398570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.300917</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.162882</td>\n",
       "      <td>0.082244</td>\n",
       "      <td>0.345012</td>\n",
       "      <td>13.311664</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>47.800100</td>\n",
       "      <td>22.605397</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>30.569839</td>\n",
       "      <td>130.592939</td>\n",
       "      <td>2.773964</td>\n",
       "      <td>7.611645</td>\n",
       "      <td>33.531139</td>\n",
       "      <td>15.295295</td>\n",
       "      <td>20.328332</td>\n",
       "      <td>118.413674</td>\n",
       "      <td>32.668651</td>\n",
       "      <td>22.805652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.203307</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.211807</td>\n",
       "      <td>0.069556</td>\n",
       "      <td>0.335723</td>\n",
       "      <td>13.785032</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>46.066403</td>\n",
       "      <td>20.853942</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>32.521534</td>\n",
       "      <td>123.678421</td>\n",
       "      <td>2.668080</td>\n",
       "      <td>8.206404</td>\n",
       "      <td>32.979449</td>\n",
       "      <td>15.033746</td>\n",
       "      <td>20.089239</td>\n",
       "      <td>121.144959</td>\n",
       "      <td>30.442728</td>\n",
       "      <td>23.403616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.256296</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.080188</td>\n",
       "      <td>0.125608</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.649623</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>45.712356</td>\n",
       "      <td>21.377179</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>27.967913</td>\n",
       "      <td>126.723619</td>\n",
       "      <td>3.286000</td>\n",
       "      <td>11.075037</td>\n",
       "      <td>37.483739</td>\n",
       "      <td>14.895510</td>\n",
       "      <td>17.376096</td>\n",
       "      <td>149.058857</td>\n",
       "      <td>32.772074</td>\n",
       "      <td>23.163784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.351086</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054925</td>\n",
       "      <td>0.075671</td>\n",
       "      <td>0.385303</td>\n",
       "      <td>8.973253</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>41.863051</td>\n",
       "      <td>22.616414</td>\n",
       "      <td>0.242189</td>\n",
       "      <td>...</td>\n",
       "      <td>27.021822</td>\n",
       "      <td>74.655933</td>\n",
       "      <td>3.736386</td>\n",
       "      <td>10.136228</td>\n",
       "      <td>37.666414</td>\n",
       "      <td>13.484046</td>\n",
       "      <td>21.128904</td>\n",
       "      <td>194.584197</td>\n",
       "      <td>32.281384</td>\n",
       "      <td>23.780587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.230624</td>\n",
       "      <td>0.035837</td>\n",
       "      <td>0.053812</td>\n",
       "      <td>0.110743</td>\n",
       "      <td>0.336408</td>\n",
       "      <td>9.336744</td>\n",
       "      <td>0.193978</td>\n",
       "      <td>43.406801</td>\n",
       "      <td>20.842553</td>\n",
       "      <td>0.212957</td>\n",
       "      <td>...</td>\n",
       "      <td>32.606337</td>\n",
       "      <td>110.171030</td>\n",
       "      <td>3.162182</td>\n",
       "      <td>11.979568</td>\n",
       "      <td>35.755054</td>\n",
       "      <td>14.319992</td>\n",
       "      <td>19.077618</td>\n",
       "      <td>153.719588</td>\n",
       "      <td>32.896717</td>\n",
       "      <td>23.078828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.362810</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.054184</td>\n",
       "      <td>0.050074</td>\n",
       "      <td>0.676265</td>\n",
       "      <td>9.562037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>42.264973</td>\n",
       "      <td>22.752890</td>\n",
       "      <td>0.207380</td>\n",
       "      <td>...</td>\n",
       "      <td>25.416376</td>\n",
       "      <td>66.817892</td>\n",
       "      <td>3.253249</td>\n",
       "      <td>10.199674</td>\n",
       "      <td>36.682645</td>\n",
       "      <td>13.111032</td>\n",
       "      <td>21.180285</td>\n",
       "      <td>207.092959</td>\n",
       "      <td>31.524216</td>\n",
       "      <td>24.633575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.350926</td>\n",
       "      <td>0.032057</td>\n",
       "      <td>0.092482</td>\n",
       "      <td>0.130880</td>\n",
       "      <td>0.358890</td>\n",
       "      <td>10.128333</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>43.736571</td>\n",
       "      <td>20.615554</td>\n",
       "      <td>0.184561</td>\n",
       "      <td>...</td>\n",
       "      <td>28.349972</td>\n",
       "      <td>100.414759</td>\n",
       "      <td>3.536064</td>\n",
       "      <td>12.396366</td>\n",
       "      <td>36.297142</td>\n",
       "      <td>13.755184</td>\n",
       "      <td>18.309939</td>\n",
       "      <td>161.319258</td>\n",
       "      <td>32.275104</td>\n",
       "      <td>22.633430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.419643</td>\n",
       "      <td>0.046901</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.098149</td>\n",
       "      <td>0.376772</td>\n",
       "      <td>10.329859</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>40.635957</td>\n",
       "      <td>21.016866</td>\n",
       "      <td>0.231078</td>\n",
       "      <td>...</td>\n",
       "      <td>22.338756</td>\n",
       "      <td>57.523884</td>\n",
       "      <td>3.284246</td>\n",
       "      <td>9.952191</td>\n",
       "      <td>36.076187</td>\n",
       "      <td>13.478254</td>\n",
       "      <td>22.231341</td>\n",
       "      <td>204.279198</td>\n",
       "      <td>29.913855</td>\n",
       "      <td>25.225166</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 22112 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name      A1BG      A1CF       A2M     A2ML1   A3GALT2     A4GALT  \\\n",
       "0          0.377997  0.035485  0.087716  0.263830  0.350411  12.057979   \n",
       "1          0.345151  0.034716  0.127103  0.137183  0.000000  13.425112   \n",
       "2          0.300917  0.000000  0.162882  0.082244  0.345012  13.311664   \n",
       "3          0.203307  0.000000  0.211807  0.069556  0.335723  13.785032   \n",
       "4          0.256296  0.000000  0.080188  0.125608  0.000000  11.649623   \n",
       "5          0.351086  0.000000  0.054925  0.075671  0.385303   8.973253   \n",
       "6          0.230624  0.035837  0.053812  0.110743  0.336408   9.336744   \n",
       "7          0.362810  0.000000  0.054184  0.050074  0.676265   9.562037   \n",
       "8          0.350926  0.032057  0.092482  0.130880  0.358890  10.128333   \n",
       "9          0.419643  0.046901  0.000000  0.098149  0.376772  10.329859   \n",
       "\n",
       "gene_name     A4GNT       AAAS       AACS     AADAC  ...     ZWILCH  \\\n",
       "0          0.319075  53.020671  20.904988  0.338778  ...  37.376160   \n",
       "1          0.000000  47.729696  21.923315  0.227967  ...  37.457048   \n",
       "2          0.000000  47.800100  22.605397  0.000000  ...  30.569839   \n",
       "3          0.000000  46.066403  20.853942  0.000000  ...  32.521534   \n",
       "4          0.000000  45.712356  21.377179  0.000000  ...  27.967913   \n",
       "5          0.000000  41.863051  22.616414  0.242189  ...  27.021822   \n",
       "6          0.193978  43.406801  20.842553  0.212957  ...  32.606337   \n",
       "7          0.000000  42.264973  22.752890  0.207380  ...  25.416376   \n",
       "8          0.000000  43.736571  20.615554  0.184561  ...  28.349972   \n",
       "9          0.000000  40.635957  21.016866  0.231078  ...  22.338756   \n",
       "\n",
       "gene_name       ZWINT      ZXDA       ZXDB       ZXDC     ZYG11A     ZYG11B  \\\n",
       "0          182.345536  3.376170   9.657581  36.781075  16.005638  18.488987   \n",
       "1          150.343398  2.518407   9.277073  33.313392  16.573176  20.549266   \n",
       "2          130.592939  2.773964   7.611645  33.531139  15.295295  20.328332   \n",
       "3          123.678421  2.668080   8.206404  32.979449  15.033746  20.089239   \n",
       "4          126.723619  3.286000  11.075037  37.483739  14.895510  17.376096   \n",
       "5           74.655933  3.736386  10.136228  37.666414  13.484046  21.128904   \n",
       "6          110.171030  3.162182  11.979568  35.755054  14.319992  19.077618   \n",
       "7           66.817892  3.253249  10.199674  36.682645  13.111032  21.180285   \n",
       "8          100.414759  3.536064  12.396366  36.297142  13.755184  18.309939   \n",
       "9           57.523884  3.284246   9.952191  36.076187  13.478254  22.231341   \n",
       "\n",
       "gene_name         ZYX      ZZEF1       ZZZ3  \n",
       "0           72.885622  33.661747  22.205224  \n",
       "1          107.455344  33.147009  23.398570  \n",
       "2          118.413674  32.668651  22.805652  \n",
       "3          121.144959  30.442728  23.403616  \n",
       "4          149.058857  32.772074  23.163784  \n",
       "5          194.584197  32.281384  23.780587  \n",
       "6          153.719588  32.896717  23.078828  \n",
       "7          207.092959  31.524216  24.633575  \n",
       "8          161.319258  32.275104  22.633430  \n",
       "9          204.279198  29.913855  25.225166  \n",
       "\n",
       "[10 rows x 22112 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plsr = PLSRegression(n_components=2)\n",
    "\n",
    "#Replace rows w/ 'TGFb-and-RA-high' with the numerical values of 'TGFb' and 'Retinoic Acid'\n",
    "def replaceTreatmentsInDF(df):\n",
    "    for treatment in sampleIDs_transform.keys():\n",
    "        sit = sampleIDs_transform[treatment]\n",
    "        df.loc[df.sampleID == treatment, ['TGFb', 'Retinoic Acid']] = sit['TGFb'], sit['retinoicAcid']\n",
    "    return df\n",
    "\n",
    "\n",
    "df_plsr = df_original.copy()\n",
    "#Only need these columns so remove everything else\n",
    "df_plsr = df_plsr[['sampleID', 'tpm', 'gene_name']]\n",
    "changeTreatmentNameInRows(df_plsr, columnName='sampleID')\n",
    "\n",
    "#Zero out NaN values\n",
    "df_plsr[['TGFb', 'Retinoic Acid']] = 0\n",
    "replaceTreatmentsInDF(df_plsr)\n",
    "\n",
    "#Group new datable by treatments and make the columns have the TPM values\n",
    "df_plsr = df_plsr.pivot_table(index=['TGFb', 'Retinoic Acid'], columns='gene_name', values='tpm', fill_value=0)\n",
    "\n",
    "#Undo the pivot table to make dataframe all columns \n",
    "df_plsr.reset_index(inplace=True)\n",
    "\n",
    "#Make a dataframe of just genes\n",
    "treatments = ['TGFb', 'Retinoic Acid']\n",
    "genes = df_plsr[df_plsr.columns.difference(treatments)]\n",
    "\n",
    "display(df_plsr)\n",
    "display(genes)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLSR Variance captured:  0.4004045357126894\n",
      "PLSR Variance captured:  0.4522931336024578\n",
      "PLSR Variance captured:  0.49686683462375353\n",
      "PLSR Variance captured:  0.5126218770309794\n",
      "PLSR Variance captured:  0.527031817796695\n",
      "PLSR Variance captured:  0.5331246230376915\n",
      "PLSR Variance captured:  0.5079130615548616\n",
      "PLSR Variance captured:  0.3615562412977845\n",
      "PLSR Variance captured:  0.32411942594593995\n"
     ]
    }
   ],
   "source": [
    "variance_capture = []\n",
    "num_genes = []\n",
    "\n",
    "\n",
    "for i in [0, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99, 0.999, 0.9999]:\n",
    "    percentile = np.quantile(genes.sum(axis = 0), i)\n",
    "    screened_genes = genes[genes.columns[genes.sum() > percentile]]\n",
    "    plsr.fit(df_plsr[treatments], screened_genes)\n",
    "    variance_score = plsr.score(df_plsr[treatments], screened_genes)\n",
    "    variance_capture.append(variance_score)\n",
    "    num_genes.append(screened_genes.shape[1])\n",
    "    print('PLSR Variance captured: ', variance_score)\n",
    "\n",
    "\n",
    "#Only percentages between 0 and 1\n",
    "plt.ylim(0, 1)\n",
    "#Make x label to show how many # of genes vs captured variance\n",
    "labels = ['%d'% i for i in num_genes]\n",
    "\n",
    "#Just to make a readable x axis plot\n",
    "genes_count = np.linspace(0, 9,len(num_genes))\n",
    "\n",
    "plt.plot(genes_count, variance_capture)\n",
    "plt.xticks(range(len(labels)), labels, rotation='vertical')\n",
    "\n",
    "plt.title('')\n",
    "plt.tight_layout()\n",
    "plt.xlabel('Top N Genes')\n",
    "plt.ylabel('Percent of Variance Captured')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLSR Variance captured:  0.40039272982335644\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>gene_name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.362298</td>\n",
       "      <td>0.029668</td>\n",
       "      <td>0.054499</td>\n",
       "      <td>0.054887</td>\n",
       "      <td>0.518592</td>\n",
       "      <td>10.323523</td>\n",
       "      <td>-0.043468</td>\n",
       "      <td>40.068333</td>\n",
       "      <td>21.136480</td>\n",
       "      <td>0.156647</td>\n",
       "      <td>...</td>\n",
       "      <td>22.720921</td>\n",
       "      <td>52.020810</td>\n",
       "      <td>3.206369</td>\n",
       "      <td>10.153712</td>\n",
       "      <td>35.357228</td>\n",
       "      <td>13.032290</td>\n",
       "      <td>21.708999</td>\n",
       "      <td>207.673088</td>\n",
       "      <td>29.807358</td>\n",
       "      <td>24.841513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.393041</td>\n",
       "      <td>0.042628</td>\n",
       "      <td>0.004124</td>\n",
       "      <td>0.067310</td>\n",
       "      <td>0.545893</td>\n",
       "      <td>8.713564</td>\n",
       "      <td>-0.029093</td>\n",
       "      <td>38.268409</td>\n",
       "      <td>20.785346</td>\n",
       "      <td>0.223194</td>\n",
       "      <td>...</td>\n",
       "      <td>20.779741</td>\n",
       "      <td>35.966035</td>\n",
       "      <td>3.508953</td>\n",
       "      <td>11.859800</td>\n",
       "      <td>36.595315</td>\n",
       "      <td>12.351237</td>\n",
       "      <td>21.193319</td>\n",
       "      <td>230.377725</td>\n",
       "      <td>30.033920</td>\n",
       "      <td>24.827794</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 22112 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "gene_name      A1BG      A1CF       A2M     A2ML1   A3GALT2     A4GALT  \\\n",
       "0          0.362298  0.029668  0.054499  0.054887  0.518592  10.323523   \n",
       "1          0.393041  0.042628  0.004124  0.067310  0.545893   8.713564   \n",
       "\n",
       "gene_name     A4GNT       AAAS       AACS     AADAC  ...     ZWILCH  \\\n",
       "0         -0.043468  40.068333  21.136480  0.156647  ...  22.720921   \n",
       "1         -0.029093  38.268409  20.785346  0.223194  ...  20.779741   \n",
       "\n",
       "gene_name      ZWINT      ZXDA       ZXDB       ZXDC     ZYG11A     ZYG11B  \\\n",
       "0          52.020810  3.206369  10.153712  35.357228  13.032290  21.708999   \n",
       "1          35.966035  3.508953  11.859800  36.595315  12.351237  21.193319   \n",
       "\n",
       "gene_name         ZYX      ZZEF1       ZZZ3  \n",
       "0          207.673088  29.807358  24.841513  \n",
       "1          230.377725  30.033920  24.827794  \n",
       "\n",
       "[2 rows x 22112 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "#Fit treatments and predict\n",
    "plsr.fit(df_plsr[treatments], genes)\n",
    "print('PLSR Variance captured: ', plsr.score(df_plsr[treatments], genes))\n",
    "printPLSRPlot(plsr.x_loadings_, plsr.y_loadings_, treatments)\n",
    "\n",
    "#Predict values of new treatment\n",
    "Y_pred = plsr.predict([[10, 400], [14, 300]])\n",
    "display(pd.DataFrame(Y_pred, columns=genes.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10, 22114)\n"
     ]
    }
   ],
   "source": [
    "import tensorflow as tf\n",
    "\n",
    "from tensorflow import keras\n",
    "from tensorflow.keras import layers\n",
    "from tensorflow.keras.layers.experimental import preprocessing\n",
    "\n",
    "print(df_plsr.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1/100\n",
      "1/1 [==============================] - 0s 433ms/step - loss: 4.1166 - accuracy: 0.0000e+00\n",
      "Epoch 2/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 64.8442 - accuracy: 0.0000e+00\n",
      "Epoch 3/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 56.1766 - accuracy: 0.0000e+00\n",
      "Epoch 4/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 30.6466 - accuracy: 0.0000e+00\n",
      "Epoch 5/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 1.9947 - accuracy: 0.0000e+00\n",
      "Epoch 6/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 26.7312 - accuracy: 0.0000e+00\n",
      "Epoch 7/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 37.0741 - accuracy: 0.0000e+00\n",
      "Epoch 8/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 36.2133 - accuracy: 0.0000e+00\n",
      "Epoch 9/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 29.1235 - accuracy: 0.0000e+00\n",
      "Epoch 10/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 18.9906 - accuracy: 0.0000e+00\n",
      "Epoch 11/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 7.6665 - accuracy: 0.0000e+00\n",
      "Epoch 12/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 4.0766 - accuracy: 0.0000e+00\n",
      "Epoch 13/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 10.0593 - accuracy: 0.0000e+00\n",
      "Epoch 14/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 11.8345 - accuracy: 0.0000e+00\n",
      "Epoch 15/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 10.7441 - accuracy: 0.0000e+00\n",
      "Epoch 16/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 7.8808 - accuracy: 0.0000e+00\n",
      "Epoch 17/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 4.0485 - accuracy: 0.0000e+00\n",
      "Epoch 18/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.2915 - accuracy: 0.0000e+00\n",
      "Epoch 19/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 2.5132 - accuracy: 0.0000e+00\n",
      "Epoch 20/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 3.2359 - accuracy: 0.0000e+00\n",
      "Epoch 21/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 2.8567 - accuracy: 0.0000e+00\n",
      "Epoch 22/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 1.6902 - accuracy: 0.0000e+00\n",
      "Epoch 23/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 0.1371 - accuracy: 0.0000e+00\n",
      "Epoch 24/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 1.0007 - accuracy: 0.0000e+00\n",
      "Epoch 25/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 1.1450 - accuracy: 0.0000e+00\n",
      "Epoch 26/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.6788 - accuracy: 0.0000e+00\n",
      "Epoch 27/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.3216 - accuracy: 0.0000e+00\n",
      "Epoch 28/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.5943 - accuracy: 0.0000e+00\n",
      "Epoch 29/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.3288 - accuracy: 0.0000e+00\n",
      "Epoch 30/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.4494 - accuracy: 0.0000e+00\n",
      "Epoch 31/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.6124 - accuracy: 0.0000e+00\n",
      "Epoch 32/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 0.3047 - accuracy: 0.0000e+00\n",
      "Epoch 33/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.4539 - accuracy: 0.0000e+00\n",
      "Epoch 34/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.6386 - accuracy: 0.0000e+00\n",
      "Epoch 35/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.3972 - accuracy: 0.0000e+00\n",
      "Epoch 36/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.2758 - accuracy: 0.0000e+00\n",
      "Epoch 37/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.4212 - accuracy: 0.0000e+00\n",
      "Epoch 38/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1676 - accuracy: 0.0000e+00\n",
      "Epoch 39/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.4873 - accuracy: 0.0000e+00\n",
      "Epoch 40/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 0.6481 - accuracy: 0.0000e+00\n",
      "Epoch 41/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.4435 - accuracy: 0.0000e+00\n",
      "Epoch 42/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1433 - accuracy: 0.0000e+00\n",
      "Epoch 43/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.2673 - accuracy: 0.0000e+00\n",
      "Epoch 44/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.0876 - accuracy: 0.0000e+00\n",
      "Epoch 45/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 0.1926 - accuracy: 0.0000e+00\n",
      "Epoch 46/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.0883 - accuracy: 0.0000e+00\n",
      "Epoch 47/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.2038 - accuracy: 0.0000e+00\n",
      "Epoch 48/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.0880 - accuracy: 0.0000e+00\n",
      "Epoch 49/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.1472 - accuracy: 0.0000e+00\n",
      "Epoch 50/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 0.1113 - accuracy: 0.0000e+00\n",
      "Epoch 51/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.0904 - accuracy: 0.0000e+00\n",
      "Epoch 52/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1374 - accuracy: 0.0000e+00\n",
      "Epoch 53/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.0871 - accuracy: 0.0000e+00\n",
      "Epoch 54/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.1160 - accuracy: 0.0000e+00\n",
      "Epoch 55/100\n",
      "1/1 [==============================] - 0s 12ms/step - loss: 0.1849 - accuracy: 0.0000e+00\n",
      "Epoch 56/100\n",
      "1/1 [==============================] - 0s 11ms/step - loss: 0.1181 - accuracy: 0.0000e+00\n",
      "Epoch 57/100\n",
      "1/1 [==============================] - 0s 44ms/step - loss: 0.2207 - accuracy: 0.0000e+00\n",
      "Epoch 58/100\n",
      "1/1 [==============================] - 0s 12ms/step - loss: 0.1664 - accuracy: 0.0000e+00\n",
      "Epoch 59/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.2337 - accuracy: 0.0000e+00\n",
      "Epoch 60/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.2504 - accuracy: 0.0000e+00\n",
      "Epoch 61/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.0875 - accuracy: 0.0000e+00\n",
      "Epoch 62/100\n",
      "1/1 [==============================] - 0s 9ms/step - loss: 0.2515 - accuracy: 0.0000e+00\n",
      "Epoch 63/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.1258 - accuracy: 0.0000e+00\n",
      "Epoch 64/100\n",
      "1/1 [==============================] - 0s 9ms/step - loss: 0.3072 - accuracy: 0.0000e+00\n",
      "Epoch 65/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.3614 - accuracy: 0.0000e+00\n",
      "Epoch 66/100\n",
      "1/1 [==============================] - ETA: 0s - loss: 0.1424 - accuracy: 0.0000e+ - 0s 3ms/step - loss: 0.1424 - accuracy: 0.0000e+00\n",
      "Epoch 67/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.3189 - accuracy: 0.0000e+00\n",
      "Epoch 68/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.4004 - accuracy: 0.0000e+00\n",
      "Epoch 69/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.2073 - accuracy: 0.0000e+00\n",
      "Epoch 70/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.2934 - accuracy: 0.0000e+00\n",
      "Epoch 71/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.4206 - accuracy: 0.0000e+00\n",
      "Epoch 72/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.2777 - accuracy: 0.0000e+00\n",
      "Epoch 73/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1688 - accuracy: 0.0000e+00\n",
      "Epoch 74/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.2568 - accuracy: 0.0000e+00\n",
      "Epoch 75/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.0902 - accuracy: 0.0000e+00\n",
      "Epoch 76/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.3545 - accuracy: 0.0000e+00\n",
      "Epoch 77/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.4468 - accuracy: 0.0000e+00\n",
      "Epoch 78/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.2888 - accuracy: 0.0000e+00\n",
      "Epoch 79/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1542 - accuracy: 0.0000e+00\n",
      "Epoch 80/100\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1/1 [==============================] - 0s 4ms/step - loss: 0.2559 - accuracy: 0.0000e+00\n",
      "Epoch 81/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1157 - accuracy: 0.0000e+00\n",
      "Epoch 82/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.3035 - accuracy: 0.0000e+00\n",
      "Epoch 83/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.3918 - accuracy: 0.0000e+00\n",
      "Epoch 84/100\n",
      "1/1 [==============================] - 0s 6ms/step - loss: 0.2463 - accuracy: 0.0000e+00\n",
      "Epoch 85/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1700 - accuracy: 0.0000e+00\n",
      "Epoch 86/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.2633 - accuracy: 0.0000e+00\n",
      "Epoch 87/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.1309 - accuracy: 0.0000e+00\n",
      "Epoch 88/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.2669 - accuracy: 0.0000e+00\n",
      "Epoch 89/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.3508 - accuracy: 0.0000e+00\n",
      "Epoch 90/100\n",
      "1/1 [==============================] - 0s 8ms/step - loss: 0.2158 - accuracy: 0.0000e+00\n",
      "Epoch 91/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1774 - accuracy: 0.0000e+00\n",
      "Epoch 92/100\n",
      "1/1 [==============================] - 0s 7ms/step - loss: 0.2637 - accuracy: 0.0000e+00\n",
      "Epoch 93/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.1384 - accuracy: 0.0000e+00\n",
      "Epoch 94/100\n",
      "1/1 [==============================] - 0s 2ms/step - loss: 0.2400 - accuracy: 0.0000e+00\n",
      "Epoch 95/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.3198 - accuracy: 0.0000e+00\n",
      "Epoch 96/100\n",
      "1/1 [==============================] - 0s 2ms/step - loss: 0.1940 - accuracy: 0.0000e+00\n",
      "Epoch 97/100\n",
      "1/1 [==============================] - 0s 5ms/step - loss: 0.1785 - accuracy: 0.0000e+00\n",
      "Epoch 98/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.2588 - accuracy: 0.0000e+00\n",
      "Epoch 99/100\n",
      "1/1 [==============================] - 0s 3ms/step - loss: 0.1401 - accuracy: 0.0000e+00\n",
      "Epoch 100/100\n",
      "1/1 [==============================] - 0s 4ms/step - loss: 0.2206 - accuracy: 0.0000e+00\n",
      "WARNING:tensorflow:11 out of the last 11 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7fd2268f9e50> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/guide/function#controlling_retracing and https://www.tensorflow.org/api_docs/python/tf/function for  more details.\n",
      "[[-0.75373036]]\n"
     ]
    }
   ],
   "source": [
    "#https://towardsdatascience.com/multi-output-model-with-tensorflow-keras-functional-api-875dd89aa7c6\n",
    "\n",
    "\n",
    "input_layers = tf.keras.Input(shape=(3, ))\n",
    "previous_layer = input_layers\n",
    "\n",
    "layer = tf.keras.layers.Dense(units=3, activation='relu')(previous_layer)\n",
    "op = tf.keras.layers.Dense(units=1)(layer)\n",
    "output_layers= [op]\n",
    "\n",
    "gene_names = ['AADAC']\n",
    "\n",
    "# for name in range(2):\n",
    "#     layer = tf.keras.layers.Dense(units=32, activation='relu')(previous_layer)\n",
    "#     output = tf.keras.layers.Dense(units=1)(layer)\n",
    "#     output_layers.append(output)\n",
    "#     previous_layer = layer\n",
    "# model = tf.keras.Model(input_layers, output_layers)\n",
    "\n",
    "model = tf.keras.Sequential([\n",
    "    layers.Dense(units='3'),\n",
    "    layers.Dense(units='64'),\n",
    "    layers.Dense(units='64'),\n",
    "    layers.Dense(units='1')\n",
    "])\n",
    "\n",
    "model.compile(\n",
    "    optimizer=tf.optimizers.Adam(learning_rate=0.01),\n",
    "    loss='mean_absolute_error',\n",
    "    metrics=[tf.keras.metrics.Accuracy()]\n",
    ")\n",
    "\n",
    "model.fit(\n",
    "    df_plsr[treatments], genes[gene_names],\n",
    "    epochs=100,\n",
    "    # suppress logging\n",
    "    #verbose=0,\n",
    "    # Calculate validation results on 20% of the training data\n",
    "    #validation_split = 0.2\n",
    ")\n",
    "\n",
    "results = model.predict([[10, 400]])\n",
    "print(results)\n",
    "\n",
    "#mse = tf.metrics.mean_squared_error(labels=genes[gene_names], predictions=results)\n",
    "\n",
    "#accuracy, update_op = tf.metrics.accuracy(labels=genes[gene_names], predictions=results)\n",
    "\n",
    "\n",
    "#display(genes[gene_names])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "pca = PCA(n_components=3)\n",
    "\n",
    "#Guys if you increase the number of PCs you get a lot more variance capture\n",
    "\n",
    "\n",
    "def printPCA_2D(new_df, title):\n",
    "    pca.fit(new_df)\n",
    "\n",
    "    print('''\n",
    "    Variance for PCs = %s \n",
    "    Sum of variances = %.03f  \n",
    "    pca.shape = %s where rows=loadings=%d, columns=PCs=%d''' \n",
    "          % (pca.explained_variance_ratio_, np.sum(pca.explained_variance_ratio_), \n",
    "             pca.components_.shape, pca.components_.shape[1], pca.components_.shape[0]))\n",
    "    plt.xlabel('PC1')\n",
    "    plt.ylabel('PC2')\n",
    "    plt.title(title)\n",
    "\n",
    "    for i in range(pca.components_.shape[1]):\n",
    "        plt.scatter(pca.components_[0, i], pca.components_[1, i], label=new_df.columns[i])\n",
    "\n",
    "    #https://stackoverflow.com/questions/7908636/possible-to-make-labels-appear-when-hovering-over-a-point-in-matplotlib\n",
    "    mplcursors.cursor(hover=True)\n",
    "\n",
    "    plt.legend(bbox_to_anchor=(1.35, 1)) \n",
    "    plt.show()\n",
    "\n",
    "    \n",
    "def printPCA_3D(new_df, title):\n",
    "    pca.fit(new_df)\n",
    "\n",
    "    print('''\n",
    "    Variance for PCs = %s \n",
    "    Sum of variances = %.03f  \n",
    "    pca.shape = %s where rows=loadings=%d, columns=PCs=%d''' \n",
    "          % (pca.explained_variance_ratio_, np.sum(pca.explained_variance_ratio_), \n",
    "             pca.components_.shape, pca.components_.shape[1], pca.components_.shape[0]))\n",
    "    \n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111, projection='3d')\n",
    "    plt.xlabel('PC1')\n",
    "    plt.ylabel('PC2')\n",
    "    plt.title(title)\n",
    "\n",
    "    for i in range(pca.components_.shape[1]):\n",
    "        ax.scatter(xs=pca.components_[0, i], ys=pca.components_[1, i], \n",
    "                       zs=pca.components_[2, i], label=new_df.columns[i])\n",
    "\n",
    "\n",
    "    plt.legend(bbox_to_anchor=(1.6, 1)) \n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def getPCA_DFs(df_pca, groupby=['sampleID', 'gene_name'], keepColumns=['sampleID', 'tpm', 'gene_name'], topN=5000):\n",
    "    \n",
    "    df_pca = df_pca[keepColumns]\n",
    "\n",
    "    #Group data by two columns, which will later be a regular row and column\n",
    "    gb = df_pca.groupby(groupby).mean().sort_values(by=['tpm'], ascending=False)[:topN]\n",
    "    \n",
    "    \n",
    "    display(gb)\n",
    "    #Extract rows and columns of top N results by indices\n",
    "    rows = set([groupby[0] for groupby in gb.index])\n",
    "    columns = set([groupby[1] for groupby in gb.index])\n",
    "\n",
    "    #Create a new dataframe\n",
    "    new_df = pd.DataFrame(0, index=rows, columns=columns)\n",
    "    new_df_log = new_df.copy()\n",
    "\n",
    "    #Iterate\n",
    "    for (row, col) in gb.index:\n",
    "        tpv_val = gb.loc[row, col].tpm\n",
    "        new_df.loc[row, col] = tpv_val\n",
    "        new_df_log.loc[row, col] = np.log(tpv_val)\n",
    "        \n",
    "    return new_df, new_df_log\n",
    "\n",
    "\n",
    "#From Meyer treatments as rows, genes as columns\n",
    "\n",
    "df_pca = df_original.copy()\n",
    "\n",
    "groupby=['sampleID', 'gene_name'] #Should only be two things! groupby[0] = row, groupby[1] = column\n",
    "changeTreatmentNameInRows(df_pca, columnName='sampleID')\n",
    "\n",
    "new_df, new_df_log = getPCA_DFs(df_pca, groupby=groupby, topN=5000)\n",
    "\n",
    "#display(new_df)\n",
    "print()\n",
    "printPCA_2D(new_df_log, title='Loadings Plot for Log-Normal PCA' )\n",
    "\n",
    "print()\n",
    "printPCA_2D(new_df, title='Loadings Plot for Normal PCA' )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
