{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: MacOSX\n",
      "Populating the interactive namespace from numpy and matplotlib\n"
     ]
    }
   ],
   "source": [
    "%pylab\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import LabelEncoder, OneHotEncoder\n",
    "from sklearn.cross_decomposition import PLSRegression\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.stats import zscore\n",
    "import mplcursors\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "DeSeqOutputAllConds = 'DeSeqOutputAllConds.tsv'\n",
    "si2_si4_RNA_seq = 'si2-si4_RNA-seq-pipeline-output-normalized.tsv'\n",
    "le = LabelEncoder()\n",
    "oh_enc = OneHotEncoder(handle_unknown='ignore')\n",
    "plsr = PLSRegression(n_components=2)\n",
    "\n",
    "\n",
    "def getTSVFile(filepath):\n",
    "    data = pd.read_csv (filepath, sep = '\\t')\n",
    "    return data\n",
    "\n",
    "#not used or necessary\n",
    "def integerEncode(dataframe, columnNametoEncode, newEncodedColumnName):\n",
    "    dataframe[columnNametoEncode] = le.fit_transform(df[newEncodedColumnName])\n",
    "    return dataframe\n",
    "  \n",
    "#https://stackoverflow.com/questions/37292872/how-can-i-one-hot-encode-in-python  \n",
    "def oneHotEncode(df, columnsToEncode):\n",
    "    \n",
    "    for column in columnsToEncode:\n",
    "        # Get one hot encoding of column and drop 'column'\n",
    "        one_hot = pd.get_dummies(df[column])\n",
    "        # Drop column  as it is now encoded\n",
    "        df = df.drop(column, axis = 1)\n",
    "        # Join the encoded df\n",
    "        df = df.join(one_hot)\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def zScoreData(df, columnsToZScore):\n",
    "    for column in columnsToZScore:\n",
    "        df[column] = zscore(df[column])\n",
    "        \n",
    "    return df\n",
    "\n",
    "def fixBadChromosomeLabeling(df, column='chrom'): \n",
    "    replace = list(range(1, 23)) + ['X', 'Y']\n",
    "    \n",
    "    for i in replace:\n",
    "        df[column].replace(str(i), 'chr' + str(i), inplace=True)\n",
    "        \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = getTSVFile(si2_si4_RNA_seq)\n",
    "df = df.loc[df['counts'] > 0]#Filter out rows where count is zero\n",
    "df = fixBadChromosomeLabeling(df)\n",
    "\n",
    "#Set up copies to play around with lader\n",
    "df_original = df.copy()\n",
    "\n",
    "#Remove unnecessary columns\n",
    "#['experiment', 'sampleID', 'counts', 'tpm', 'gene_name', 'chrom'] Leon..KEEP\n",
    "df = df[['sampleID', 'tpm', 'gene_name']]\n",
    "\n",
    "#display(df.sampleID.unique())\n",
    "\n",
    "df = oneHotEncode(df, ['sampleID'])\n",
    "df = zScoreData(df, ['tpm'])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#Leon if you want to get top n%\n",
    "# Probably simpler - https://pandas.pydata.org/pandas-docs/version/0.17.0/generated/pandas.DataFrame.nlargest.html\n",
    "# Sort, .head(...) - https://stackoverflow.com/questions/53405458/selecting-the-top-50-percentage-names-from-the-columns-of-a-pandas-dataframe\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_original' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-51802826fa4f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_original\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mdf_test\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf_test\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdf_test\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'counts'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf_test\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene_name'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_original' is not defined"
     ]
    }
   ],
   "source": [
    "df_test = df_original.copy()\n",
    "df_test = df_test.loc[df_test['counts'] > 0]\n",
    "\n",
    "\n",
    "print(len(df_test['gene_name'].unique()))\n",
    "\n",
    "#Group by gene and therapeutic level, sorting by average tpm value w/ highest tpm first, \n",
    "#getting top 20\n",
    "\n",
    "#print(len(df_test[['gene_name'== 'MT-CO2', 'sampleID' =='35-TGFb-and-RA-high']]))\n",
    "df_test.query('gene_name==\"MT-CO2\" & sampleID==\"35-TGFb-and-RA-high\"')\n",
    "\n",
    "\n",
    "\n",
    "#Homework: Get top 10 results, for gene_name, sampleID, and both and PLOT IT\n",
    "df_test.groupby(['gene_name', 'sampleID' ]).mean().sort_values(by=['tpm'], ascending=False)[:20]\n",
    "\n",
    "\n",
    "#1. One hot encode the genes. When they are numbers they wont disappear using 'groupby'\n",
    "#2. Groupby by sampleID(treatments)\n",
    "#3. Remove columns [counts, rpm, rpkm, tx_start, tx_end, TSS_loc]\n",
    "#4. Transpose the dataframe so the genes are the row\n",
    "#5. Sort by tpm and get top n results. So you top 20 genes\n",
    "#6. Transpose it back and now you have top 20 columns\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retinoic Acid (50, 200 and 400 nM)\n",
    "# TGF-Î² (1.25, 5, and 10 ng/mL)\n",
    "\n",
    "\n",
    "sampleIDs_transform = {\n",
    "  'RA-high': {'TGFb': 0, 'retinoicAcid': 400, 'EtOH': 0}, \n",
    "  'EtOH-nlDensity': {'TGFb': 0, 'retinoicAcid': 0, 'EtOH': 0},\n",
    "  'RA-med': {'TGFb': 0, 'retinoicAcid': 200, 'EtOH': 0}, \n",
    "  'EtOH-halfDensity': {'TGFb': 0, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'RA-low': {'TGFb': 0, 'retinoicAcid': 50, 'EtOH': 0},\n",
    "  'EtOH-highDensity': {'TGFb': 0, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'TGFb-and-RA-low': {'TGFb': 1.25, 'retinoicAcid': 50, 'EtOH': 0}, \n",
    "  'TGFb-and-RA-high': {'TGFb': 10, 'retinoicAcid': 400, 'EtOH': 0},\n",
    "  'TGFb-and-RA-med': {'TGFb': 5, 'retinoicAcid': 200, 'EtOH': 0}, \n",
    "  'TGFb-high': {'TGFb': 10, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'TGFb-low': {'TGFb': 1.25, 'retinoicAcid': 0, 'EtOH': 0}, \n",
    "  'TGFb-med': {'TGFb': 5, 'retinoicAcid': 0, 'EtOH': 0}\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#https://stackoverflow.com/questions/11346283/renaming-columns-in-pandas\n",
    "def combineTreatmentColumns(df, exceptions):\n",
    "    df.rename(columns=lambda x: x[3:] if x not in exceptions else x, inplace=True)\n",
    "    return df\n",
    "\n",
    "def changeTreatmentNameInRows(df, columnName='sampleID', exceptions={}):\n",
    "    df[columnName] = df[columnName].apply(lambda x: x[3:] if x not in exceptions else x)\n",
    "    return df\n",
    "    \n",
    "    \n",
    "def printPLSRPlot(x_loadings_, y_loadings_, loadings):\n",
    "    plt.scatter(y_loadings_[:, 0], y_loadings_[:, 1], color='blue', label='tpm')\n",
    "    \n",
    "    for i in range(x_loadings_.shape[0]):\n",
    "        plt.scatter(x_loadings_[i,0], x_loadings_[i, 1], label=loadings[i])\n",
    "        \n",
    "    plt.legend(bbox_to_anchor=(1.53, 1))\n",
    "    plt.xlabel('PC1')\n",
    "    plt.ylabel('PC2')\n",
    "    plt.show()\n",
    "    \n",
    "def getPLSR_DFs(groupby=['sampleID', 'gene_name'], columns=['sampleID', 'tpm', 'gene_name'], topN=5000):\n",
    "    df_plsr = df_original.copy()\n",
    "    df_plsr = df_plsr[columns]\n",
    "\n",
    "    gb = df_plsr.groupby(groupby).mean().sort_values(by=['tpm'], ascending=False)[:topN]\n",
    "    #Extract rows and columns of top N results by indices\n",
    "    rows = set([groupby[0] for groupby in gb.index])\n",
    "    columns = set([groupby[1] for groupby in gb.index])\n",
    "\n",
    "    #Create a new dataframe\n",
    "    new_df = pd.DataFrame(0, index=rows, columns=columns)\n",
    "    new_df_log = new_df.copy()\n",
    "\n",
    "    #Iterate\n",
    "    for (row, col) in gb.index:\n",
    "        tpv_val = gb.loc[row, col].tpm\n",
    "        new_df.loc[row, col] = tpv_val\n",
    "        new_df_log.loc[row, col] = np.log(tpv_val)\n",
    "        \n",
    "    return new_df, new_df_log\n",
    "\n",
    "\n",
    "#new_df, new_df_log = getPCA_DFs()\n",
    "\n",
    "#Something wrong my PLSR logic\n",
    "\n",
    "#display(df_original)\n",
    "\n",
    "\n",
    "def createPLSRMatrix(df, newColumns=['TGFb', 'retinoicAcid', 'EtOH']):\n",
    "    #Create a new dataframe    \n",
    "    new_indexes = [gene for (gene, treatment) in df.index]\n",
    "    new_df = pd.DataFrame(0, index=new_indexes, columns=newColumns)\n",
    "    \n",
    "    test = df[:100]\n",
    "        \n",
    "    for i in range(len(new_df.index)):\n",
    "        \n",
    "        (gene_name, treatment) = df.iloc[i].name\n",
    "        \n",
    "        \n",
    "        new_df.iat[i, 0] = sampleIDs_transform[treatment]['TGFb']\n",
    "        new_df.iat[i, 1]= sampleIDs_transform[treatment]['retinoicAcid']\n",
    "        new_df.iat[i, 2]= sampleIDs_transform[treatment]['EtOH']\n",
    "        \n",
    "    \n",
    "\n",
    "    return new_df\n",
    "\n",
    "\n",
    "df_plsr = df_original.copy()\n",
    "df_plsr = df_plsr[['sampleID', 'tpm', 'gene_name']]\n",
    "\n",
    "\n",
    "\n",
    "#df_plsr = oneHotEncode(df_plsr, ['sampleID'])\n",
    "#df_plsr = combineTreatmentColumns(df_plsr, {'tpm', 'gene_name'})\n",
    "#display(df_plsr.columns) Shows the fixed column names\n",
    "\n",
    "\n",
    "#display(df_plsr[df_plsr['sampleID'].str.contains(\"TGFb-and-RA-low\")])\n",
    "\n",
    "changeTreatmentNameInRows(df_plsr, columnName='sampleID')\n",
    "\n",
    "\n",
    "#display(df_plsr[df_plsr['sampleID'] == 'RA-high'])\n",
    "\n",
    "\n",
    "df_plsr = df_plsr.groupby(['gene_name', 'sampleID']).mean().sort_values(by=['tpm'], ascending=False)[:50]\n",
    "Y = df_plsr['tpm']\n",
    "\n",
    "\n",
    "\n",
    "better_df = createPLSRMatrix(df_plsr)\n",
    "display(df_plsr)\n",
    "display(better_df)\n",
    "display(Y)\n",
    "\n",
    "#display(better_df, Y[:20])\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "loadings = [col for col in df_plsr.columns if col != 'tpm'] #Remove 'tpm' to fit for PLSR\n",
    "plsr.fit(better_df, Y)\n",
    "print('PLSR Variance captured: ', plsr.score(better_df, Y))\n",
    "printPLSRPlot(plsr.x_loadings_, plsr.y_loadings_, better_df.columns)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>tpm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35-TGFb-and-RA-high</th>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>40095.437187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31-TGFb-high</th>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>37089.321282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14-EtOH-highDensity</th>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>34573.327005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20-EtOH-halfDensity</th>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>30770.942552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36-RA-low</th>\n",
       "      <th>MT-CO2</th>\n",
       "      <td>30730.360023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01-TGFb-low</th>\n",
       "      <th>ACTB</th>\n",
       "      <td>5924.099692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30-TGFb-med</th>\n",
       "      <th>KRT8</th>\n",
       "      <td>5919.391663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29-TGFb-and-RA-low</th>\n",
       "      <th>ACTB</th>\n",
       "      <td>5862.717586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51-RA-med</th>\n",
       "      <th>ACTB</th>\n",
       "      <td>5857.966945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33-TGFb-low</th>\n",
       "      <th>ACTB</th>\n",
       "      <td>5830.268826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500 rows Ã 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        tpm\n",
       "sampleID            gene_name              \n",
       "35-TGFb-and-RA-high MT-CO2     40095.437187\n",
       "31-TGFb-high        MT-CO2     37089.321282\n",
       "14-EtOH-highDensity MT-CO2     34573.327005\n",
       "20-EtOH-halfDensity MT-CO2     30770.942552\n",
       "36-RA-low           MT-CO2     30730.360023\n",
       "...                                     ...\n",
       "01-TGFb-low         ACTB        5924.099692\n",
       "30-TGFb-med         KRT8        5919.391663\n",
       "29-TGFb-and-RA-low  ACTB        5862.717586\n",
       "51-RA-med           ACTB        5857.966945\n",
       "33-TGFb-low         ACTB        5830.268826\n",
       "\n",
       "[500 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "    Variance for PCs = [0.78440204 0.21156343 0.0026108 ] \n",
      "    Sum of variances = 0.999  \n",
      "    pca.shape = (3, 15) where rows=loadings=15, columns=PCs=3\n",
      "\n",
      "    Variance for PCs = [0.60421674 0.17176473 0.09608637] \n",
      "    Sum of variances = 0.872  \n",
      "    pca.shape = (3, 15) where rows=loadings=15, columns=PCs=3\n"
     ]
    }
   ],
   "source": [
    "from sklearn.decomposition import PCA\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "\n",
    "pca = PCA(n_components=3)\n",
    "\n",
    "#Guys if you increase the number of PCs you get a lot more variance capture\n",
    "\n",
    "\n",
    "def printPCA_2D(new_df, title):\n",
    "    pca.fit(new_df)\n",
    "\n",
    "    print('''\n",
    "    Variance for PCs = %s \n",
    "    Sum of variances = %.03f  \n",
    "    pca.shape = %s where rows=loadings=%d, columns=PCs=%d''' \n",
    "          % (pca.explained_variance_ratio_, np.sum(pca.explained_variance_ratio_), \n",
    "             pca.components_.shape, pca.components_.shape[1], pca.components_.shape[0]))\n",
    "    plt.xlabel('PC1')\n",
    "    plt.ylabel('PC2')\n",
    "    plt.title(title)\n",
    "\n",
    "    for i in range(pca.components_.shape[1]):\n",
    "        plt.scatter(pca.components_[0, i], pca.components_[1, i], label=new_df.columns[i])\n",
    "\n",
    "    mplcursors.cursor(hover=True)\n",
    "\n",
    "    plt.legend(bbox_to_anchor=(1.35, 1)) \n",
    "    plt.show()\n",
    "\n",
    "    \n",
    "def printPCA_3D(new_df, title):\n",
    "    pca.fit(new_df)\n",
    "\n",
    "    print('''\n",
    "    Variance for PCs = %s \n",
    "    Sum of variances = %.03f  \n",
    "    pca.shape = %s where rows=loadings=%d, columns=PCs=%d''' \n",
    "          % (pca.explained_variance_ratio_, np.sum(pca.explained_variance_ratio_), \n",
    "             pca.components_.shape, pca.components_.shape[1], pca.components_.shape[0]))\n",
    "    \n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(111, projection='3d')\n",
    "    plt.xlabel('PC1')\n",
    "    plt.ylabel('PC2')\n",
    "    plt.title(title)\n",
    "\n",
    "    for i in range(pca.components_.shape[1]):\n",
    "        ax.scatter(xs=pca.components_[0, i], ys=pca.components_[1, i], \n",
    "                       zs=pca.components_[2, i], label=new_df.columns[i])\n",
    "\n",
    "\n",
    "    plt.legend(bbox_to_anchor=(1.6, 1)) \n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def getPCA_DFs(groupby=['sampleID', 'gene_name'], keepColumns=['sampleID', 'tpm', 'gene_name'], topN=5000):\n",
    "    df_pca = df_original.copy()\n",
    "    df_pca = df_pca[keepColumns]\n",
    "\n",
    "    #Group data by two columns, which will later be a regular row and column\n",
    "    gb = df_pca.groupby(groupby).mean().sort_values(by=['tpm'], ascending=False)[:topN]\n",
    "    \n",
    "    \n",
    "    display(gb)\n",
    "    #Extract rows and columns of top N results by indices\n",
    "    rows = set([groupby[0] for groupby in gb.index])\n",
    "    columns = set([groupby[1] for groupby in gb.index])\n",
    "\n",
    "    #Create a new dataframe\n",
    "    new_df = pd.DataFrame(0, index=rows, columns=columns)\n",
    "    new_df_log = new_df.copy()\n",
    "\n",
    "    #Iterate\n",
    "    for (row, col) in gb.index:\n",
    "        tpv_val = gb.loc[row, col].tpm\n",
    "        new_df.loc[row, col] = tpv_val\n",
    "        new_df_log.loc[row, col] = np.log(tpv_val)\n",
    "        \n",
    "    return new_df, new_df_log\n",
    "\n",
    "#From Meyer treatments as rows, genes as columns\n",
    "\n",
    "\n",
    "groupby=['sampleID', 'gene_name'] #Should only be two things! groupby[0] = row, groupby[1] = column\n",
    "new_df, new_df_log = getPCA_DFs(groupby=groupby, topN=500)\n",
    "\n",
    "#display(new_df)\n",
    "\n",
    "printPCA_2D(new_df_log, title='Loadings Plot for Log-Normal PCA' )\n",
    "printPCA_2D(new_df, title='Loadings Plot for Normal PCA' )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
